src/main/java/org/tugraz/sysds/runtime/instructions/fed/AggregateBinaryFEDInstruction.java

```diff
@@ -81,95 +86,295 @@ public void processInstruction(ExecutionContext ec) {
+				int resultColumn = c + (!matrixVectorOp ? beginDims[1] : 0);
+				resultBlock.quickSetValue(resultRow,
+						resultColumn,
+						resultBlock.quickGetValue(resultRow, resultColumn) + mb.quickGetValue(r, c));
+			}
+	}
+	
+	private static Future<FederatedResponse> executeMVMultiply(FederatedRange range, FederatedData fedData,
+			MatrixBlock vector, boolean matrixVectorOp) {
+		if (!fedData.isInitialized()) {
+			throw new DMLRuntimeException("Not all FederatedData was initialized for federated matrix");
+		}
+		int[] beginDimsInt = range.getBeginDimsInt();
+		int[] endDimsInt = range.getEndDimsInt();
+		// params for federated request
+		List<Object> params = new ArrayList<>();
+		// we broadcast the needed part of the small vector
+		MatrixBlock vectorSlice;
+		if (!matrixVectorOp) {
+			// if we the size already is ok, we do not have to copy a slice
+			int length = endDimsInt[0] - beginDimsInt[0];
+			if (vector.getNumColumns() == length) {
+				vectorSlice = vector;
+			}
+			else {
+				vectorSlice = new MatrixBlock(1, length, false);
+				vector.slice(0, 0, beginDimsInt[0], endDimsInt[0] - 1, vectorSlice);
+			}
+		}
+		else {
+			int length = endDimsInt[1] - beginDimsInt[1];
+			if (vector.getNumRows() == length) {
+				vectorSlice = vector;
+			}
+			else {
+				vectorSlice = new MatrixBlock(length, 1, false);
+				vector.slice(beginDimsInt[1], endDimsInt[1] - 1, 0, 0, vectorSlice);
+			}
+		}
+		params.add(vectorSlice);
+		params.add(matrixVectorOp); // if is matrix vector multiplication true, otherwise false
+		return fedData
+				.executeFederatedOperation(new FederatedRequest(FederatedRequest.FedMethod.MATVECMULT, params), true);
+	}
+	
+	private static class FederatedMMTask implements Callable<Void> {
+		private FederatedRange _range;
+		private FederatedData _data;
+		private MutablePair<FederatedRange, MatrixBlock> _result;
+		private MatrixBlock _otherMatrix;
+		private boolean _distributeCols;
+		
+		public FederatedMMTask(FederatedRange range, FederatedData fedData,
+				MutablePair<FederatedRange, MatrixBlock> result, MatrixBlock otherMatrix, boolean distributeCols) {
+			_range = range;
+			_data = fedData;
+			_result = result;
+			_otherMatrix = otherMatrix;
+			_distributeCols = distributeCols;
+		}
+		
+		@Override
+		public Void call() throws Exception {
+			if (_distributeCols) {
+				executeColWiseMVMultiplication();
+			}
+			else {
+				executeRowWiseVMMultiplications();
+			}
+			return null;
+		}
+		
+		/**
+		 * Distribute the non or smaller federated block as row vectors to the federated worker and do row number of
+		 * times a vector-matrix multiplication. Non or smaller federated block is left operand.
+		 *
+		 * @throws InterruptedException if .get() on federated response future fails -> interrupted
+		 * @throws ExecutionException   if .get() on federated response future fails -> execution failed
+		 */
+		private void executeRowWiseVMMultiplications() throws InterruptedException, ExecutionException {
```

Comment:
We need to talk about this with @mboehm7. along with the col Wise Multiplication.